File: app/src/main/java/com/ashwinrao/sanbox/viewmodel/ViewModelFactory.java
package com.ashwinrao.sanbox.viewmodel;


import androidx.annotation.NonNull;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

import com.ashwinrao.sanbox.data.repo.BoxRepository;
import com.ashwinrao.sanbox.data.repo.ItemRepository;

import javax.inject.Inject;
import javax.inject.Singleton;

@Singleton
public class ViewModelFactory implements ViewModelProvider.Factory {

    private final BoxRepository boxRepo;
    private final ItemRepository itemRepo;

    @Inject
    public ViewModelFactory(@NonNull BoxRepository boxRepo, @NonNull ItemRepository itemRepo) {
        this.boxRepo = boxRepo;
        this.itemRepo = itemRepo;
    }

    @SuppressWarnings("unchecked")
    @NonNull
    @Override
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
        if (modelClass.isAssignableFrom(BoxViewModel.class)) {
            return (T) new BoxViewModel(boxRepo);
        } else if (modelClass.isAssignableFrom(CategoryViewModel.class)) {
            return (T) new CategoryViewModel();
        } else if (modelClass.isAssignableFrom(ItemViewModel.class)) {
            return (T) new ItemViewModel(itemRepo);
        } else
            throw new IllegalArgumentException("ViewModel could not be loaded");
    }
}

File: app/src/main/java/com/ashwinrao/packup/viewmodel/ViewModelFactory.kt
package com.ashwinrao.packup.viewmodel


import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import com.ashwinrao.packup.data.AppDatabase

import com.ashwinrao.packup.data.repo.BoxRepository
import com.ashwinrao.packup.data.repo.ItemRepository

import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class ViewModelFactory @Inject
constructor(private val database: AppDatabase,
            private val boxRepo: BoxRepository,
            private val itemRepo: ItemRepository) : ViewModelProvider.Factory {

    @Suppress("UNCHECKED_CAST")
    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        return when {
            modelClass.isAssignableFrom(BoxViewModel::class.java) -> BoxViewModel(boxRepo) as T
            modelClass.isAssignableFrom(CategoryViewModel::class.java) -> CategoryViewModel() as T
            modelClass.isAssignableFrom(ItemViewModel::class.java) -> ItemViewModel(itemRepo) as T
            modelClass.isAssignableFrom(InsertionViewModel::class.java) -> InsertionViewModel(database) as T
            else -> throw IllegalArgumentException("ViewModel could not be loaded")
        }
    }
}
